Distinct offline sync function (send all queued preparation pt1)

dron22 opened the pull request:
This PR adds a distinct function `offline.Sync()` for offline activity syncing. This is a first step, to separate offline syncing from the heartbeats pipeline. It is a preparation to sync all queued heartbeats on wakatime cli execution.

- Adds helper functions `popHeartbeats` and `pushHeartbeats` to pop and push without keeping the db connection open.

Related to: https://github.com/wakatime/wakatime-cli/issues/449

dron22 committed:
Add offline.Sync() function

Will make exactly one request to the WakaTime API including as many
heartbeats, as defined by sync-limit parameter.

dron22 committed:
Rename api client method

dron22 committed:
Use sync function in offline sync command

dron22 committed:
Do not consider offline-disabled parameter in offline sync command

dron22 committed:
Optionally use queue filepath from parameter in offlinesync

alanhamlett commented on pkg/offline/offline.go:
```diff
@@ -157,6 +157,146 @@ func WithQueue(filepath string, syncLimit int) (heartbeat.HandleOption, error) {
+	return err
+}
+
+func popHeartbeats(filepath string, syncLimit int) ([]heartbeat.Heartbeat, error) {
```
This is great!

alanhamlett commented on pkg/offline/offline.go:
```diff
@@ -157,6 +157,146 @@ func WithQueue(filepath string, syncLimit int) (heartbeat.HandleOption, error) {
+		return fmt.Errorf("failed initialize new queue: %s", err)
+	}
+
+	err = queue.PushMany(hh)
```
Can we have some retry here? I'm thinking if there are any errors due to another process having the global boltdb lock, we should retry so heartbeats aren't lost.

dron22 replied:
Added a retry. Check new commits :+1:

alanhamlett commented on pkg/offline/offline.go:
```diff
@@ -157,6 +157,146 @@ func WithQueue(filepath string, syncLimit int) (heartbeat.HandleOption, error) {
+	if len(withInvalidStatus) > 0 {
+		log.Debugf("pushing %d heartbeat(s) with invalid result to queue", len(withInvalidStatus))
+
+		err = pushHeartbeats(filepath, withInvalidStatus)
```
Maybe the retry goes here instead of in `pushHeartbeats`, up to you.

alanhamlett replied:
`syncLimit` should just be `limit` because this code has nothing to do with syncing.. it's just popping multiple (up to 24) heartbeats from the offline db.

dron22 replied:
Added a retry. Check new commits :+1:

alanhamlett commented on pkg/offline/offline.go:
```diff
@@ -157,6 +157,146 @@ func WithQueue(filepath string, syncLimit int) (heartbeat.HandleOption, error) {
+	return err
+}
+
+func popHeartbeats(filepath string, syncLimit int) ([]heartbeat.Heartbeat, error) {
```
`syncLimit` should never be larger than `24`, but it looks like this comes directly from user input?

alanhamlett replied:
`syncLimit` should just be `limit` because this code has nothing to do with syncing.. it's just popping multiple (up to 24) heartbeats from the offline db.

dron22 replied:
Alright, after reading https://github.com/wakatime/wakatime-cli/pull/455#discussion_r650560600, I see the issue. I will add a constant `sendLimit` with value `24` here. Then I will add correct application of sync limit in https://github.com/wakatime/wakatime-cli/pull/455.

So what should be the default sync limit, in case `--sync-offline-activity` parameter is not provided?

alanhamlett replied:
Default should be 1k.

dron22 replied:
Alright, will adjust this in https://github.com/wakatime/wakatime-cli/pull/455

alanhamlett commented on pkg/offline/offline.go:
```diff
@@ -157,6 +157,146 @@ func WithQueue(filepath string, syncLimit int) (heartbeat.HandleOption, error) {
 }
 
+// Sync returns a function to send queued heartbeats to the WakaTime API.
+func Sync(filepath string, syncLimit int) func(next heartbeat.Handle) error {
```
`syncLimit` should just be `limit` because this code has nothing to do with syncing.. it's just popping multiple (up to 24) heartbeats from the offline db.

dron22 committed:
Add retry for offline sync requeuing

alanhamlett commented on pkg/offline/offline.go:
```diff
@@ -157,6 +167,164 @@ func WithQueue(filepath string, syncLimit int) (heartbeat.HandleOption, error) {
+
+				requeueErr := pushHeartbeats(filepath, hh)
+				if requeueErr != nil {
+					log.Errorf("failed to push heatbeats after api error to queue: %s", requeueErr)
```
This should be a debug log.

dron22 replied:
I guess we're best here with `warning`. I mean something really went wrong here, and also if we want to recover lost heartbeats, we don't want to depend on the user setting log level to debug, right?

alanhamlett replied:
We've already got the error logging at `if count >= maxRequeueAttempts`, so we don't need to log two errors for the same thing. Here we can still retry, so the heartbeat might not be lost. If we retry, nothing should be printed unless `--verbose` was passed. That's because it's highly possible two wakatime-cli processes try to write at the same time, with one not able to acquire the global lock. That's not a problem, as long as we don't reach the `maxRequeueAttempts`.

dron22 replied:
Yeah, this is a good idea. Adjusted following your remarks :+1:

alanhamlett commented on pkg/offline/offline.go:
```diff
@@ -157,6 +167,164 @@ func WithQueue(filepath string, syncLimit int) (heartbeat.HandleOption, error) {
+
+			for {
+				if count >= maxRequeueAttempts {
+					log.Errorf("abort requeuing after %d unsuccessful attempts", count)
```
Let's change this to something like:

```go
log.Errorf("failed writing heartbeat to offline queue: %s", string(json.Marshal(hh)[0]))
```

That code probably doesn't work, but shows what we want is to log the heartbeat to `~/.wakatime.log` when unable to save to the offline db. That allows parsing the log file to recover lost heartbeats.

gandarez commented on pkg/offline/offline.go:
```diff
@@ -157,6 +167,169 @@ func WithQueue(filepath string, syncLimit int) (heartbeat.HandleOption, error) {
+	// handle leftover heartbeats
+	leftovers := len(hh) - len(results)
+	if leftovers > 0 {
+		log.Warnf("Missing %d results from api.", leftovers)
```
```suggestion
		log.Warnf("missing %d results from api", leftovers)
```

dron22 replied:
sharp eyes! :+1:

dron22 replied:
adjusted. see new commit below.

gandarez replied:
Don't know why linter didn't catch it

gandarez reviewed (commented):
one small remark

dron22 committed:
Move retry logic to distinct function

In second part of the send all queued preparation, this will also
be used in WithQueue function.
